libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentType.java

```diff
@@ -113,9 +115,26 @@ public XContent xContent() {
         }
     };
 
-    public static final MediaTypeParser<XContentType> mediaTypeParser = new MediaTypeParser<>(XContentType.values(),
-        Map.of("application/*", JSON, "application/x-ndjson", JSON));
-
+    private static final Pattern VERSION_PATTERN = Pattern.compile("\\d+");
+    private static final String COMPATIBLE_WITH_PARAMETER_NAME = "compatible-with";
+     public static final MediaTypeParser<XContentType> mediaTypeParser = new MediaTypeParser.Builder<XContentType>()
+        .withMediaTypeAndParams("application/smile", SMILE, Collections.emptyMap())
+        .withMediaTypeAndParams("application/cbor", CBOR, Collections.emptyMap())
+        .withMediaTypeAndParams("application/json", JSON, Map.of("charset", Pattern.compile("UTF-8")))
+        .withMediaTypeAndParams("application/yaml", YAML, Map.of("charset", Pattern.compile("UTF-8")))
```

Comment:
I think this is technically non-passive (only supporting UTF-8 encoding) and should probably be mentioned in the breaking changes ... thus only targeting 8.0 and (outside of this PR) ensure that a 7.x server maps the compatible header to standard header ... i.e. (application/vnd.elasticsearch+yaml; compatible-with=7 == application/yaml)

Reply:
that's a good point. I think clients will have to upgrade to latest 7.x client to get `compatible-with`  sent on their request.
 We should also support that in 7.x server. Will raise an issue/pr

Reply:
handling of versioned media types in a form `application/vnd.elasticsearch+json;compatible-with=7` will be handled in this pr https://github.com/elastic/elasticsearch/pull/63071